ddtrace/internal/wrapping/context.py

```diff
@@ -0,0 +1,686 @@
+            pop_top
+        _except:
+        """
+    )
+
+    CONTEXT_RETURN.parse(
+        r"""
+            load_const                  {context}
+            load_method                 $__return__
+            rot_three
+            rot_three
+            call_method                 1
+        """
+    )
+
+    CONTEXT_FOOT.parse(
+        r"""
+            with_except_start
+            pop_top
+            reraise                     1
+        """
+    )
+
+elif sys.version_info >= (3, 9):
+    CONTEXT_HEAD.parse(
+        r"""
+            load_const                  {context}
+            setup_with                  @_except
+            pop_top
+        _except:
+        """
+    )
+
+    CONTEXT_RETURN.parse(
+        r"""
+            load_const                  {context}
+            load_method                 $__return__
+            rot_three
+            rot_three
+            call_method                 1
+        """
+    )
+
+    CONTEXT_FOOT.parse(
+        r"""
+            with_except_start
+            pop_top
+            reraise
+        """
+    )
+
+
+elif sys.version_info >= (3, 7):
+    CONTEXT_HEAD.parse(
+        r"""
+            load_const                  {context}
+            setup_with                  @_except
+            pop_top
+        _except:
+        """
+    )
+
+    CONTEXT_RETURN.parse(
+        r"""
+            load_const                  {context}
+            load_method                 $__return__
+            rot_three
+            rot_three
+            call_method                 1
+        """
+    )
+
+    CONTEXT_FOOT.parse(
+        r"""
+            with_cleanup_start
+            with_cleanup_finish
+            end_finally
+            load_const                  None
+            return_value
+        """
+    )
+
+
+# This is abstract and should not be used directly
+class BaseWrappingContext(t.ContextManager):
+    __priority__: int = 0
+    __frame__: t.Optional[FrameType] = None
+
+    def __init__(self, f: FunctionType):
+        self.__wrapped__ = f
+
+    @classmethod
+    def wrapped(cls, f: FunctionType) -> "BaseWrappingContext":
+        if cls.is_wrapped(f):
+            context = cls.extract(f)
+            assert isinstance(context, cls)  # nosec
+        else:
+            context = cls(f)
+            context.wrap()
+        return context
+
+    def __return__(self, value):
+        return value
+
+    @classmethod
+    def is_wrapped(cls, _f: FunctionType) -> bool:
+        raise NotImplementedError
+
+    @classmethod
+    def extract(cls, _f: FunctionType) -> "BaseWrappingContext":
+        raise NotImplementedError
+
+    def wrap(self) -> None:
+        raise NotImplementedError
+
+    def unwrap(self) -> None:
+        raise NotImplementedError
+
+
+# This is the public interface exported by this module
+class WrappingContext(BaseWrappingContext):
+    @override
+    def __enter__(self, _frame: FrameType) -> "WrappingContext":
```

Comment:
should we also provide a base implementation of `__exit__`.

`AbstractContextManager` makes you implement it even though there is a base implementation which is just `return None`, we could do the same here?

If we don't, then we enforce that people acknowledge and handle errors, which isn't a bad thing?

Reply:
We could but I thought it wouldn't make much sense to use a wrapping context manager if you didn't have a need to implement `__exit__`/`__return__`. So in this sense, I'd view them as mandatory to implement by any subclass